### PR TITLE
feat: add 'range_id' algorithm for 'request-id' plugin

### PR DESCRIPTION
**Pull request by jiangfucheng**
### Description

Fixes # ([issue](https://github.com/apache/apisix/issues/8720))

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/plugins/request-id.lua [hunk 1]
```diff
@@ -40,7 +40,9 @@ local schema = {
     properties = {
         header_name = {type = "string", default = "X-Request-Id"},
         include_in_response = {type = "boolean", default = true},
-        algorithm = {type = "string", enum = {"uuid", "snowflake", "nanoid"}, default = "uuid"}
+        algorithm = {type = "string", enum = {"uuid", "snowflake", "nanoid", "random_id"}, default = "uuid"},
+        random_id_length = { type = "integer",  minimum = 6, default = 16 },
```
Comment (spacewander) [inline apisix/plugins/request-id.lua:44]: The name "random" is confusing. It doesn't show why this change can satisfy the requirement of opentelemetry in https://github.com/apache/apisix/issues/8720. Is it because the other algorithms are not random?

Comment (jiangfucheng) [reply apisix/plugins/request-id.lua:44]: In https://github.com/apache/apisix/issues/8720 issue, the user needs a random id that can satisfy the `[0-9a-f]{32}` regular expression. I think that the character content and id length of other algorithms cannot meet the requirements. In this "random_id" algorithm, users only need to set random_id_strs="abcdef0123456789" and random_id_length=32 to meet the requirements.

Comment (spacewander) [reply apisix/plugins/request-id.lua:51]: As I explained, the name "random" is confusing. Most of the other algorithms also contain a part which is generated randomly.

Comment (jiangfucheng) [reply apisix/plugins/request-id.lua:51]: Do you have some good suggestions?

**Comment by jiangfucheng** [comment]
@spacewander Can you help restart ci task? I fixed some error, thanks.

File: apisix/plugins/request-id.lua [hunk 2]
```diff
@@ -202,14 +215,28 @@ local function next_id()
...
 
+-- generate random id
+local function get_random_id(strs, length)
+    local res = ""
+    for i = 1,length do
+        local idx = math_random(str_len(strs))
+        res = res .. str_sub(strs, idx, idx)
+    end
+    return res
+end
```
Comment (spacewander) [inline apisix/plugins/request-id.lua:226]: I provide a faster implementation:
```
local function get_random_id(strs, length)
    local res = ffi.new("unsigned char[?]", length)
    for i = 0, length - 1 do
        res[i] = string.byte(strs, math_random(#strs))
    end
    return ffi.string(res, length)
end
```

You can write a benchmark to see the difference.

File: docs/en/latest/plugins/request-id.md [hunk 3]
```diff
@@ -44,7 +44,9 @@ The Plugin will not add a unique ID if the request already has a header with the
 | ------------------- | ------- | -------- | -------------- | ------------------------------- | ---------------------------------------------------------------------- |
 | header_name         | string  | False    | "X-Request-Id" |                                 | Header name for the unique request ID.                                 |
 | include_in_response | boolean | False    | true           |                                 | When set to `true`, adds the unique request ID in the response header. |
-| algorithm           | string  | False    | "uuid"         | ["uuid", "snowflake", "nanoid"] | Algorithm to use for generating the unique request ID.                 |
+| algorithm           | string  | False    | "uuid"         | ["uuid", "snowflake", "nanoid", "specified_character_id"] | Algorithm to use for generating the unique request ID.                 |
+| specified_character_id_strs      | string | 否 | "abcdefghijklmnopqrstuvwxyzABCDEFGHIGKLMNOPQRSTUVWXYZ0123456789| The minimum string length is 6 | Character set for specified_character_id |
```
[the remark this answers is not in the document]
Comment (jiangfucheng) [reply docs/en/latest/plugins/request-id.md]: Fixed

File: apisix/plugins/request-id.lua [hunk 4]
```diff
@@ -40,7 +43,17 @@ local schema = {
     properties = {
         header_name = {type = "string", default = "X-Request-Id"},
         include_in_response = {type = "boolean", default = true},
-        algorithm = {type = "string", enum = {"uuid", "snowflake", "nanoid"}, default = "uuid"}
+        algorithm = {
+            type = "string",
+            enum = {"uuid", "snowflake", "nanoid", "specified_character_id"},
+            default = "uuid"
+        },
+        specified_character_id_length = { type = "integer",  minimum = 6, default = 16 },
```
[the remark this answers is not in the document]
Comment (jiangfucheng) [reply apisix/plugins/request-id.lua]: Updated

File: apisix/plugins/request-id.lua [hunk 5]
```diff
@@ -40,7 +43,17 @@ local schema = {
...
+        specified_character_id_strs = {
```
[the remark this answers is not in the document]
Comment (jiangfucheng) [reply apisix/plugins/request-id.lua]: It's good idea, updated.

File: apisix/plugins/request-id.lua [hunk 6]
```diff
@@ -40,7 +43,17 @@ local schema = {
...
+            type = "string",
+            minLength = 6,
```
[the remark this answers is not in the document]
Comment (jiangfucheng) [reply apisix/plugins/request-id.lua]: I just want to avoid too short length that users set, too short id may repeat.

File: apisix/plugins/request-id.lua [hunk 7]
```diff
@@ -40,7 +43,26 @@ local schema = {
...
+            type = "object",
+            properties = {
+                length = {
+                    type = "integer",
+                    minimum = 6,
+                    default = 16
+                },
+                char_set = {
+                    type = "string",
+                    minLength = 6,
```
Comment (spacewander) [inline apisix/plugins/request-id.lua:61]: Let's add a comment about why this number is chosen

File: apisix/plugins/request-id.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (spacewander) [reply apisix/plugins/request-id.lua]: Actually I mean `strs` to `range`. 
But as you decide to name it range_id, it's fine to me.

File: t/plugin/request-id2.t [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+use t::APISIX 'no_plan';
+
+log_level('debug');
```
[the remark this answers is not in the document]
Comment (jiangfucheng) [reply t/plugin/request-id2.t]: removed

File: t/plugin/request-id2.t [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+                    )
+                    if not res then
+                        ngx.log(ngx.ERR, err)
+                        return
+                    end
+                    local id = res.headers["X-Request-Id"]
+                    if not id then
+                        return -- ignore if the data is not synced yet.
+                    end
+                    if string.len(id) ~= 16 then
```
[the remark this answers is not in the document]
Comment (jiangfucheng) [reply t/plugin/request-id2.t]: Fixed

File: t/plugin/request-id2.t [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                        ngx.say(id)
+                        ngx.say("incorrect length for id")
+                        return
+                    end
```
[the remark this answers is not in the document]
Comment (jiangfucheng) [reply t/plugin/request-id2.t:159]: Done.

File: apisix/plugins/request-id.lua [hunk 11]
```diff
@@ -40,7 +43,27 @@ local schema = {
...
+            type = "object",
+            properties = {
+                length = {
+                    type = "integer",
+                    minimum = 6,
+                    default = 16
+                },
+                char_set = {
+                    type = "string",
+                    -- The Length is set to  6 just avoid too short length, it may repeat
```
Comment (spacewander) [inline apisix/plugins/request-id.lua:61]:
```suggestion
                    -- The Length is set to 6 just avoid too short length, it may repeat
```

File: t/plugin/request-id2.t [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+                    local id = res.headers["X-Request-Id"]
+                    if not id then
+                        return -- ignore if the data is not synced yet.
+                    end
+                    if #id ~= 16 then
+                        ngx.say(id)
+                        ngx.say("incorrect length for id")
+                        return
+                    end
+                    for j = 180, #id do
```
Comment (spacewander) [inline t/plugin/request-id2.t:160]: The content check code is confusing. Could you explain how it works?

Comment (jiangfucheng) [reply t/plugin/request-id2.t]: `for j = 1, #id do` The var `j` should start from 1. At first, i want use `[a-zA-Z0-9]{16}` regex expression to check the id if correct, but Lua seems don't support `{}` syntax， so i use for loop to check each  character.

Comment (jiangfucheng) [reply t/plugin/request-id2.t]: But this may be a stupid way, if you have better one, I will improve it.

Comment (spacewander) [reply t/plugin/request-id2.t]: There is a huge improvement space for the current O(n^2) solution. Please think about a better one.

Comment (jiangfucheng) [reply t/plugin/request-id2.t]: I updated the test.

**Review by spacewander — APPROVED** [review]

**Comment by jiangfucheng** [comment]
@soulbird @monkeyDluffy6017  Could help me review this PR? Thanks.

**Review by soulbird — APPROVED** [review]
